File: src/main/Java/boj_step13/boj18870.java
import java.io.*;
import java.util.Arrays;
import java.util.HashMap;

public class boj18870 {
    public static void main(String[] args) throws IOException {
        BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter writer = new BufferedWriter(new OutputStreamWriter(System.out));
        int n = Integer.parseInt(reader.readLine()), value = 0;
        int[] vertical = Arrays.stream(reader.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
        int[] sortedVertical = Arrays.copyOf(vertical, vertical.length);
        HashMap<Integer, Integer> map = new HashMap<>();
        Arrays.sort(sortedVertical);
        for (int key : sortedVertical)
            if (!map.containsKey(key)) map.put(key, value++);
        for (int it : vertical) writer.append(map.get(it) + " ");
        writer.close();
    }
}

File: src/main/kotlin/boj_step13/boj18870.kt
import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.InputStreamReader
import java.io.OutputStreamWriter

fun main() {
    val r = BufferedReader(InputStreamReader(System.`in`))
    val w = BufferedWriter(OutputStreamWriter(System.out))
    val n = r.readLine().toInt()
    val xArray = r.readLine().split(" ").map { it.toInt() }
    val arr = xArray.sorted().distinct()
    xArray.forEachIndexed() { index, x ->
        w.write("${arr.binarySearch(xArray[index])} ")
    }
    w.close()
}